File: Queue/c/LinkedQueue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node Node;

struct Node {
    int data;
    Node *next;
};

int size(); 
int isEmpty();
void peek();
void enqueue(int data);
void dequeue();
void print();

int count = 0;
Node* first;
Node* last;

int main() {
    enqueue(1);
    enqueue(2);
    enqueue(4);
    peek();
    print();
    dequeue();
    dequeue();
    print();
    return 0; 
}

int size() {
    return count;
}

int isEmpty() {
    return (size() == 0);
}

void peek() {
    if (isEmpty()) {
        printf("Queue kosong\n");
    } else {
        printf("Yang paling depan adalah %d\n", first->data);
    }
}

void enqueue(int data) {
    printf("Menambahkan %d ke queue\n", data);
    if (isEmpty()) {
        first = malloc(sizeof(Node));
        first->data = data;
        first->next = NULL;
        last = first;
    } else {
        Node* temp = malloc(sizeof(Node));
        temp->data = data;
        temp->next = NULL;
        last->next = temp;
        last = temp;
    }
    count++;
}

void dequeue() {
    if (isEmpty()) {
        printf("Queue kosong\n");
    } else {
        printf("Data yang dihapus adalah %d\n", first->data);
        first = first->next;
        count--;
    }
}

void print() {
    Node* temp = first;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

File: Data Structures/java/queue/linked/LinkedQueue.java
package queue.linked;

import queue.Queue;
import queue.QueueFullException;
import queue.QueueEmptyException;

public class LinkedQueue implements Queue {
    private Node first;
    private Node last;
    private int count;

    public LinkedQueue() {
        first = null;
        last = null;
        count = 0;
    }

    public LinkedQueue(Object obj) {
        first = new Node(obj);
        last = first;
        count = 1;
    }
    
    public int size() {
        return count;
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public boolean isFull() {
        return false;
    }

    public Object peek() throws QueueEmptyException {
        if (isEmpty()) {
            throw new QueueEmptyException("Queue is empty");
        }
        return first.data;
    }

    public void enqueue(Object obj) throws QueueFullException {
        if (isFull()) {
            throw new QueueFullException("Queue is full");
        }
        Node curr = new Node(obj); 
        if (isEmpty()) {
            first = curr;
        } else {
            last.next = curr;
        }
        last = curr;
        count++;
    }

    public Object dequeue() throws QueueEmptyException {
        if (isEmpty()) {
            throw new QueueEmptyException("Queue is empty");
        }
        Object obj = first.data;
        first = first.next;
        count--;
        return obj;
    }

    public void Print() {
        if (isEmpty()) {
            System.out.println("Empty.");
            return;
        }
        Node temp = first;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
    }
}
